File: trueforce/device.py
"""
trueforce.device
================
USB device open/close and init-sequence helpers.
"""

import struct
import time

import usb.core
import usb.util

from .protocol import (
    VENDOR_ID, PRODUCT_ID, EP_OUT, EP_IN,
    CMD05_SLOTS, SETUP_CMDS, SHUTDOWN_PKT,
    _hdr, _pkt,
)


class DeviceNotFoundError(RuntimeError):
    pass


def open_device(vendor_id: int = VENDOR_ID, product_id: int = PRODUCT_ID):
    """
    Find the wheel, detach kernel drivers, and set USB configuration.

    Returns
    -------
    (dev, detached_interfaces)
    """
    dev = usb.core.find(idVendor=vendor_id, idProduct=product_id)
    if dev is None:
        raise DeviceNotFoundError(
            f"Device {vendor_id:#06x}:{product_id:#06x} not found. "
            "Is the wheel connected? Do you have USB permissions?"
        )
    print(f"Device: Bus {dev.bus}  Addr {dev.address}")
    detached = []
    for cfg in dev:
        for intf in cfg:
            n = intf.bInterfaceNumber
            try:
                if dev.is_kernel_driver_active(n):
                    dev.detach_kernel_driver(n)
                    detached.append(n)
                    print(f"  Detached interface {n}")
            except usb.core.USBError as e:
                print(f"  Interface {n}: {e}")
    dev.set_configuration()
    return dev, detached


def close_device(dev, detached: list[int]) -> None:
    """Send shutdown packet and reattach kernel drivers."""
    try:
        dev.write(EP_OUT, SHUTDOWN_PKT, timeout=200)
    except Exception:
        pass
    for n in detached:
        try:
            dev.attach_kernel_driver(n)
        except Exception:
            pass
    print("Wheel disabled, interfaces reattached.")


def send_init(dev, skip_setup: bool = True) -> None:
    """
    Send the CMD=05 slot-configuration sequence and optional setup commands.

    Parameters
    ----------
    skip_setup : If True (default), skip SETUP_CMDS that cause wheel rotation.
                 Set to False only if you need full kinesthetic FFB support.
    """
    # Reset
    try:
        dev.write(EP_OUT, SHUTDOWN_PKT, timeout=200)
        time.sleep(0.05)
    except Exception:
        pass

    print("--- CMD=05 init ---")
    ok = 0
    for slot, stype in CMD05_SLOTS:
        pkt = _pkt(_hdr(0x05, slot), f"{stype:02x}00 00000000 00000000")
        dev.write(EP_OUT, pkt, timeout=100)
        try:
            resp = bytes(dev.read(EP_IN, 64, timeout=50))
            if len(resp) >= 12:
                f_val = struct.unpack_from("<f", resp, 8)[0]
                print(f"  Slot 0x{slot:02x}  param={f_val:.4f}")
            ok += 1
        except usb.core.USBError:
            print(f"  Slot 0x{slot:02x}  (no response)")
    print(f"  → {ok}/{len(CMD05_SLOTS)} OK")

    if skip_setup:
        print("--- Setup commands skipped ---")
        return

    print("--- Setup commands ---")
    for i, pkt in enumerate(SETUP_CMDS):
        dev.write(EP_OUT, pkt, timeout=100)
        print(f"  [{i}] CMD=0x{pkt[4]:02x} slot=0x{pkt[5]:02x}")
        time.sleep(0.003)
    time.sleep(0.05)
    print("  → Setup OK")

File: trueforce/device.cpp
#include "device.hpp"
#include <cstring>
#include <cstdio>
#include <thread>
#include <chrono>

namespace trueforce {

using namespace std::chrono_literals;

// ── Device::close ──────────────────────────────────────────────────

void Device::close() {
    if (!handle) return;

    // Send shutdown packet
    uint8_t buf[64];
    std::memcpy(buf, SHUTDOWN_PKT.data(), 64);
    int transferred = 0;
    libusb_bulk_transfer(handle, EP_OUT, buf, 64, &transferred, 200);

    // Reattach kernel drivers
    for (int n : detached) {
        libusb_attach_kernel_driver(handle, n);
    }
    detached.clear();

    libusb_close(handle);
    handle = nullptr;

    if (ctx) {
        libusb_exit(ctx);
        ctx = nullptr;
    }

    std::printf("Wheel disabled, interfaces reattached.\n");
}

// ── open_device ────────────────────────────────────────────────────

Device open_device(uint16_t vid, uint16_t pid) {
    Device dev;

    if (int r = libusb_init(&dev.ctx); r < 0)
        throw USBError("libusb_init failed", r);

    dev.handle = libusb_open_device_with_vid_pid(dev.ctx, vid, pid);
    if (!dev.handle) {
        libusb_exit(dev.ctx);
        dev.ctx = nullptr;
        throw DeviceNotFoundError(
            "Device not found. Is the wheel connected? Do you have USB permissions?");
    }

    // Get device info
    libusb_device* d = libusb_get_device(dev.handle);
    std::printf("Device: Bus %d  Addr %d\n",
        libusb_get_bus_number(d),
        libusb_get_device_address(d));

    // Detach kernel drivers from all interfaces
    libusb_config_descriptor* cfg = nullptr;
    libusb_get_active_config_descriptor(d, &cfg);
    if (cfg) {
        for (int i = 0; i < cfg->bNumInterfaces; ++i) {
            if (libusb_kernel_driver_active(dev.handle, i) == 1) {
                if (int r = libusb_detach_kernel_driver(dev.handle, i); r == 0) {
                    dev.detached.push_back(i);
                    std::printf("  Detached interface %d\n", i);
                } else {
                    std::printf("  Interface %d: %s\n", i, libusb_strerror((libusb_error)r));
                }
            }
        }
        libusb_free_config_descriptor(cfg);
    }

    if (int r = libusb_set_configuration(dev.handle, 1); r < 0)
        throw USBError("set_configuration failed", r);

    return dev;
}

// ── send / recv ────────────────────────────────────────────────────

void send(Device& dev, const Packet& pkt, unsigned timeout_ms) {
    uint8_t buf[64];
    std::memcpy(buf, pkt.data(), 64);
    int transferred = 0;
    libusb_bulk_transfer(dev.handle, EP_OUT, buf, 64, &transferred, timeout_ms);
}

std::vector<uint8_t> recv(Device& dev, unsigned timeout_ms) {
    uint8_t buf[64] = {};
    int transferred = 0;
    int r = libusb_bulk_transfer(dev.handle, EP_IN, buf, 64, &transferred, timeout_ms);
    if (r == 0 && transferred > 0)
        return std::vector<uint8_t>(buf, buf + transferred);
    return {};
}

// ── send_init ──────────────────────────────────────────────────────

void send_init(Device& dev, bool skip_setup) {
    // Reset
    send(dev, SHUTDOWN_PKT, 200);
    std::this_thread::sleep_for(50ms);

    std::printf("--- CMD=05 init ---\n");
    int ok = 0;
    for (const auto& [slot, stype] : CMD05_SLOTS) {
        Packet p{};
        p[0] = 0x01;
        p[4] = 0x05;
        p[5] = slot;
        p[6] = stype;
        send(dev, p, 100);

        auto resp = recv(dev, 50);
        if (resp.size() >= 12) {
            float f;
            std::memcpy(&f, resp.data() + 8, 4);
            std::printf("  Slot 0x%02x  param=%.4f\n", slot, f);
        } else {
            std::printf("  Slot 0x%02x  (no response)\n", slot);
        }
        ++ok;
    }
    std::printf("  → %d/%d OK\n", ok, (int)CMD05_SLOTS.size());

    if (skip_setup) {
        std::printf("--- Setup commands skipped ---\n");
        return;
    }

    std::printf("--- Setup commands ---\n");
    for (int i = 0; i < (int)SETUP_CMDS.size(); ++i) {
        const auto& pkt = SETUP_CMDS[i];
        send(dev, pkt, 100);
        std::printf("  [%d] CMD=0x%02x slot=0x%02x\n", i, pkt[4], pkt[5]);
        std::this_thread::sleep_for(3ms);
    }
    std::this_thread::sleep_for(50ms);
    std::printf("  → Setup OK\n");
}

} // namespace trueforce
